[FIX] use max_nb_eater defined in the share

victor-champonnois opened the pull request:
## Description
Use the max number of eater defined in the share rather than the fixed amount


## Odoo task (if applicable)
https://gestion.coopiteasy.be/web#id=8305&view_type=form&model=project.task&action=475&active_id=224

victor-champonnois committed:
[FIX] use max_nb_eater defined in the share

victor-champonnois committed:
fixup! [FIX] use max_nb_eater defined in the share

robinkeunen commented on beesdoo_pos/models/beesdoo_pos.py:
```diff
@@ -4,13 +4,20 @@
         for i, eater in enumerate(self.child_eater_ids):
             eaters[i] = eater.name
-        return tuple(eaters)
+        return eaters
```
If you're adding `@api.multi`, you should either make sure your function works with a recordset of any length or add `self.ensure_one` to the beginning of the function.

victor-champonnois replied:
fixed, thx !

victor-champonnois committed:
fixup! [FIX] use max_nb_eater defined in the share

victor-champonnois commented:
@robinkeunen Do you think the JS part is OK ? I did it a bit wildly !

robinkeunen reviewed (approved):
I can't really comment on JS quality : I don't know it. The logic seems sound :-)

huguesdk commented on beesdoo_pos/models/beesdoo_pos.py:
```diff
@@ -4,13 +4,19 @@
-        eater1, eater2, eater3 = self._get_eater()
-        return eater1, eater2, eater3
+    def _get_max_nb_eater_allowed(self):
+        if self.eater == "eater":
+            max_nb_eater_allowed = (
+                self.parent_eater_id._cooperator_share_type().max_nb_eater_allowed
+            )
+        else:
+            max_nb_eater_allowed = self._cooperator_share_type().max_nb_eater_allowed
```
a comment would be nice. i had to go look into `beesdoo_base` to understand what the `else` case is. something like:
```python
# this partner is an eater, not a cooperator, so we need to access the cooperator partner to which it is linked to to access the share type.
```

victor-champonnois replied:
done thanks :)

huguesdk commented on beesdoo_pos/models/beesdoo_pos.py:
```diff
@@ -4,13 +4,19 @@
         for i, eater in enumerate(self.child_eater_ids):
             eaters[i] = eater.name
-        return tuple(eaters)
+        return eaters
```
why was it a tuple before and not anymore?

huguesdk commented on beesdoo_pos/static/src/js/beesdoo.js:
```diff
@@ -39,24 +39,14 @@ odoo.define("beesdoo_pos.screens", function (require) {
+                            self,
+                            ".customer-delegate".concat(i + 1),
+                            result[i],
+                            "Eater".concat(i + 1, ": ")
```
there is a space missing after `Eater` (also on line 89).

huguesdk replied:
also, does this work with translations?

huguesdk reviewed (changes requested):
the js part looks good to me, but [xml template](https://github.com/beescoop/Obeesdoo/blob/12.0-use-max_nb_eater-defined-in-the-share/beesdoo_pos/static/src/xml/templates.xml) has 3 `customer-delegate`s hard-coded (several times), so it won’t work if the number is higher than 3. the template should use a loop too.

huguesdk commented on beesdoo_pos/models/beesdoo_pos.py:
```diff
@@ -4,13 +4,19 @@
+    @api.multi
+    def get_eater(self):
+        self.ensure_one()
+        eaters = [False] * self._get_max_nb_eater_allowed()
         for i, eater in enumerate(self.child_eater_ids):
             eaters[i] = eater.name
```
a comment i wrote but forgot to add appparently:

another way to write this:
```python
eaters = [eater.name for eater in self.child_eater_ids)
# pad with False
eaters.extend([False] * (self._get_max_nb_eater_allowed() - len(eaters))
```
not sure which one is more readable, though. 😅

huguesdk reviewed (commented):
since the number of eaters is variable, isn’t there a way to have a dynamic list of elements in a template without having to fiddle with number suffixes to identify them?

victor-champonnois committed:
[IMP] allow for n eaters in pos

victor-champonnois commented:
> since the number of eaters is variable, isn’t there a way to have a dynamic list of elements in a template without having to fiddle with number suffixes to identify them?

@huguesdk that's a very good idea! I refactored the whole thing, which also corrected for some issues you mentionned. 
- now get_eaters returns a list, since the number of eater is variable (the list is mutable so it seems like a better option)
- I fixed the other templates
-  I removed the "Eater" strings, which indeed where not translated. I'll ask the analyst if it's ok to remove them.

I noticed that the check for the max number of eaters is done [here](https://github.com/beescoop/Obeesdoo/blob/12.0/beesdoo_base/models/partner.py#L62), and it should also be refactored. I'll see with the analyst first.

huguesdk commented on beesdoo_pos/models/beesdoo_pos.py:
```diff
@@ -4,13 +4,11 @@
-        eater1, eater2, eater3 = self._get_eater()
-        return eater1, eater2, eater3
+        self.ensure_one()
+        eaters = []
+        # todo check for max eater
+        for eater in self.child_eater_ids:
+            eaters.append(eater.name)
+        return eaters
```
a shorter way of writing this:

```suggestion
        return [eater.name for eater in self.child_eater_ids]
```

victor-champonnois replied:
Nice thanks :)

huguesdk commented on beesdoo_pos/static/src/js/beesdoo.js:
```diff
@@ -5,14 +5,11 @@ odoo.define("beesdoo_pos.screens", function (require) {
-        }
+    var set_customer_info = function (parent_class, value) {
+        var parent = this.$(parent_class);
+        var info_div = $("<div></div>").text(value);
```
this seems like a strange way to create an element. i don’t know what the best way is, but here is another way:
```suggestion
        var info_div = document.createElement("div");
        info_div.textContent = value;
```
and then no need for the `[0]` below.

victor-champonnois replied:
Thx, I've read that the way I did this is the standard way with jquery ([see here ](https://www.w3schools.com/jquery/jquery_dom_add.asp)for instance), but your way is cleaner :)

huguesdk replied:
thanks for the information. the jquery way is ok for me (although there is no `[0]` in the example). in any case, a comment explaining what is being done would be nice to have, since it’s not obvious. something like: `create a new div containing value`.

victor-champonnois replied:
Sorry I was not clear, I used your way in the new fixup commit :)

huguesdk commented on beesdoo_pos/static/src/js/beesdoo.js:
```diff
@@ -91,23 +77,8 @@ odoo.define("beesdoo_pos.screens", function (require) {
-                        ".customer-delegate3",
-                        result[2],
-                        "Eater 3: "
+                    result.forEach((element) =>
+                        set_customer_info.call(self, ".customer-delegates", element)
```
i wouldn’t call the variable `element` because it implies that it is an html element, while it is the name of person (as a string), if i understand correctly. `customer_name`, maybe?

didn’t you lost the “Eater <n>: ” prefix? is this wanted?

victor-champonnois replied:
- OK, I went for `client_name` instead, to be consistent with the `get_client` method.

- Yes I lost the "Eater 1:" prefix, I am waiting for the analyst's answer to see if it's OK or what would be a better output.

huguesdk reviewed (changes requested):
great, much better! generic code, clearer and shorter. congrats!

victor-champonnois committed:
fixup! [IMP] allow for n eaters in pos

victor-champonnois commented:
@huguesdk Thx for the re-review :) I accounted for all your comment in the code.

huguesdk reviewed (approved):
lgtm 👍
